File: Java/백준/Gold/2636. 치즈/치즈.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

class Main {

  private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
  private static final StringBuilder sb = new StringBuilder();
  private static StringTokenizer st;

  private static int n, m;
  private static int[][] map;
  private static int remainCheese, lastCheese;
  private static final int[] dys = {0, 0, -1, 1}, dxs = {-1, 1, 0, 0};

  public static void main(String[] args) throws Exception {
    init();
    solve();
  }

  private static void init() throws Exception {
    st = new StringTokenizer(br.readLine());
    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());

    remainCheese = 0;
    map = new int[n][m];
    for (int i = 0; i < n; i++) {
      st = new StringTokenizer(br.readLine());
      for (int j = 0; j < m; j++) {
        map[i][j] = Integer.parseInt(st.nextToken());
        if (map[i][j] == 1) {
          remainCheese++;
        }
      }
    }

    br.close();
  }

  private static void solve() throws IOException {

    int time = 0;

    while (remainCheese > 0) {
      lastCheese = remainCheese;
      bfs();
      time++;
    }

    System.out.printf("%d\n%d", time, lastCheese);
  }

  private static void bfs() {
    boolean[][] visited = new boolean[n][m];
    visited[0][0] = true;
    Deque<int[]> dq = new ArrayDeque<>();
    Deque<int[]> meltingCheese = new ArrayDeque<>();
    dq.offer(new int[]{0, 0});

    while (!dq.isEmpty()) {
      int[] curr = dq.poll();
      int currY = curr[0];
      int currX = curr[1];

      for (int d = 0; d < 4; d++) {
        int nxtY = currY + dys[d];
        int nxtX = currX + dxs[d];

        // 범위를 넘지 않고, 아직 방문하지 않은 곳이라면 -> 탐색
        if (inRange(nxtY, nxtX) && !visited[nxtY][nxtX]) {
          visited[nxtY][nxtX] = true; // 방문 처리
          if (map[nxtY][nxtX] == 0) {
            dq.offer(new int[]{nxtY, nxtX}); // 다음 방문할 위치 저장
          } else {
            meltingCheese.offer(new int[]{nxtY, nxtX}); // 녹을 치즈 위치 저장
          }
        }
      }
      melting(meltingCheese);
    }
  }

  private static void melting(Deque<int[]> meltingCheese) {
    remainCheese -= meltingCheese.size();
    while (!meltingCheese.isEmpty()) {
      int[] cheese = meltingCheese.poll();
      map[cheese[0]][cheese[1]] = 0;
    }
  }

  private static boolean inRange(int y, int x) {
    return 0 <= y && y < n && 0 <= x && x < m;
  }

}

File: Python/백준/Gold/2638. 치즈/치즈.py
from sys import stdin
from collections import deque

input = stdin.readline


def in_range(y, x):
    return 0 <= y < n and 0 <= x < m


def bfs(y, x):
    dq = deque([(y, x)])
    dys, dxs = [-1, 1, 0, 0], [0, 0, -1, 1]

    visited = [[0 for _ in range(m)] for _ in range(n)]
    visited[y][x] = 1  # 방문 처리
    cheese = deque()
    flag = False
    while dq:

        cur_y, cur_x = dq.popleft()

        for dy, dx in zip(dys, dxs):
            nxt_y, nxt_x = cur_y + dy, cur_x + dx
            if in_range(nxt_y, nxt_x):
                if visited[nxt_y][nxt_x] == 0 and MAP[nxt_y][nxt_x] == 0:
                    dq.append((nxt_y, nxt_x))
                    visited[nxt_y][nxt_x] = 1

                if MAP[nxt_y][nxt_x] == 1:
                    visited[nxt_y][nxt_x] += 1
                    if visited[nxt_y][nxt_x] >= 2:
                        cheese.append((nxt_y, nxt_x))

    if cheese:
        flag = True
        for cy, cx in cheese:
            MAP[cy][cx] = 0

    return 1 if flag else 0


n, m = map(int, input().split())
MAP = [list(map(int, input().split())) for _ in range(n)]

time = 0
while True:
    v = bfs(0, 0)
    if v == 0:
        print(time)
        break

    time += v
